File: data-mesh-taxonomy-py/taxonomy/excel_service.py
"""Excel import/export, faithful to the original Node/ExcelJS behavior.

* Export emits one row per branch, preserving empty branches.
* Import validates the column structure, then processes every row inside the
  caller's single transaction: blank rows are skipped, per-row validation
  problems are recorded and skipped, and an unexpected error propagates so the
  caller rolls the whole import back (no partial corruption).
"""
from __future__ import annotations

from datetime import datetime
from io import BytesIO
from typing import Optional

from openpyxl import Workbook, load_workbook
from openpyxl.styles import Font

from .service import TaxonomyService
from .types import Level
from .validation import ValidationError, normalize_technical_owner, same_name, validate_create

SHEET = "Taxonomy"

COLUMNS = [
    "Domain Group Name",
    "Domain Group Description",
    "Domain Group Business Owner",
    "Domain Group Technical Owner",
    "Domain Name",
    "Domain Description",
    "Domain Business Owner",
    "Domain Technical Owner",
    "Subdomain Name",
    "Subdomain Description",
    "Subdomain Business Owner",
    "Subdomain Technical Owner",
    "Data Product Name",
    "Data Product Description",
    "Data Product Business Owner",
    "Data Product Technical Owner",
]

# "Domain Group Technical Owner" is optional (DG tech owner isn't required).
REQUIRED_COLUMNS = [c for c in COLUMNS if c != "Domain Group Technical Owner"]

EXAMPLE_ROW = [
    "Customer", "Customer-facing data assets", "Jane Doe", "",
    "Loyalty", "Loyalty program data", "Jane Doe", "Raj Patel",
    "Membership", "Member profiles and tiers", "Jane Doe", "Raj Patel",
    "Member Profile", "Curated member profile product", "Jane Doe", "Raj Patel",
]


class RowError(Exception):
    pass


def _cell_text(value: object) -> str:
    if value is None:
        return ""
    if isinstance(value, datetime):
        return value.isoformat()
    if isinstance(value, float) and value.is_integer():
        return str(int(value))
    return str(value).strip()


class ExcelService:
    def __init__(self, service: TaxonomyService) -> None:
        self.service = service

    # -- export ----------------------------------------------------------
    def _style(self, ws) -> None:  # noqa: ANN001
        for cell in ws[1]:
            cell.font = Font(bold=True)
        ws.freeze_panes = "A2"
        for i, name in enumerate(COLUMNS, start=1):
            ws.column_dimensions[ws.cell(row=1, column=i).column_letter].width = max(18, len(name) + 2)

    def export(self) -> bytes:
        wb = Workbook()
        ws = wb.active
        ws.title = SHEET
        ws.append(COLUMNS)

        for g in self.service.get_tree():
            g_cells = [g["name"], g["description"], g["businessOwner"], g["technicalOwner"] or ""]
            if not g["children"]:
                ws.append(g_cells + [""] * 12)
                continue
            for d in g["children"]:
                d_cells = [d["name"], d["description"], d["businessOwner"], d["technicalOwner"] or ""]
                if not d["children"]:
                    ws.append(g_cells + d_cells + [""] * 8)
                    continue
                for s in d["children"]:
                    s_cells = [s["name"], s["description"], s["businessOwner"], s["technicalOwner"] or ""]
                    if not s["children"]:
                        ws.append(g_cells + d_cells + s_cells + [""] * 4)
                        continue
                    for p in s["children"]:
                        ws.append(
                            g_cells + d_cells + s_cells
                            + [p["name"], p["description"], p["businessOwner"], p["technicalOwner"] or ""]
                        )

        self._style(ws)
        buf = BytesIO()
        wb.save(buf)
        return buf.getvalue()

    def template(self) -> bytes:
        wb = Workbook()
        ws = wb.active
        ws.title = SHEET
        ws.append(COLUMNS)
        ws.append(EXAMPLE_ROW)
        self._style(ws)
        buf = BytesIO()
        wb.save(buf)
        return buf.getvalue()

    # -- import ----------------------------------------------------------
    def import_workbook(self, buffer: bytes, actor: str = "system") -> dict:
        wb = load_workbook(BytesIO(buffer), data_only=True)
        ws = wb[SHEET] if SHEET in wb.sheetnames else wb.worksheets[0]
        if ws is None:
            raise ValidationError({"file": "Workbook contains no worksheets"})

        header_index: dict[str, int] = {}
        for col, cell in enumerate(ws[1], start=1):
            key = _cell_text(cell.value).lower()
            if key:
                header_index[key] = col

        missing = [c for c in REQUIRED_COLUMNS if c.lower() not in header_index]
        if missing:
            raise ValidationError({"file": f"Missing required column(s): {', '.join(missing)}"})

        def get(row: tuple, name: str) -> str:
            col = header_index.get(name.lower())
            return _cell_text(row[col - 1]) if col else ""

        summary = {"created": 0, "updated": 0, "skipped": 0, "errors": []}

        for row_number, row in enumerate(ws.iter_rows(min_row=2, values_only=True), start=2):
            parsed = _parse_row(row, get)
            if not any(parsed.values()):
                summary["skipped"] += 1
                continue
            try:
                _validate_row_shape(parsed)
                for level, payload in (
                    (Level.DOMAIN_GROUP, parsed["dg"]),
                    (Level.DOMAIN, parsed["domain"]),
                    (Level.SUBDOMAIN, parsed["sub"]),
                    (Level.DATA_PRODUCT, parsed["product"]),
                ):
                    if payload:
                        validate_create(level, payload)

                gid = self._upsert(Level.DOMAIN_GROUP, None, parsed["dg"], summary, actor)
                if not parsed["domain"]:
                    continue
                did = self._upsert(Level.DOMAIN, gid, parsed["domain"], summary, actor)
                if not parsed["sub"]:
                    continue
                sid = self._upsert(Level.SUBDOMAIN, did, parsed["sub"], summary, actor)
                if not parsed["product"]:
                    continue
                self._upsert(Level.DATA_PRODUCT, sid, parsed["product"], summary, actor)
            except ValidationError as err:
                summary["errors"].append({"row": row_number, "message": "; ".join(err.fields.values())})
            except RowError as err:
                summary["errors"].append({"row": row_number, "message": str(err)})

        return summary

    def _upsert(self, level: Level, parent_id: Optional[str], payload: dict, summary: dict, actor: str = "system") -> str:
        tech = normalize_technical_owner(payload.get("technicalOwner"))
        name = payload["name"]
        existing = next(
            (n for n in self.service.list(level, parent_id) if same_name(n["name"], name)), None
        )
        if existing:
            changed = (
                existing["description"] != payload["description"].strip()
                or existing["businessOwner"] != payload["businessOwner"].strip()
                or (existing["technicalOwner"] or None) != tech
                or existing["name"] != name.strip()
            )
            if changed:
                self.service.update(
                    level,
                    existing["id"],
                    {
                        "name": name,
                        "description": payload["description"],
                        "businessOwner": payload["businessOwner"],
                        "technicalOwner": tech,
                    },
                    actor=actor,
                )
                summary["updated"] += 1
            return existing["id"]

        created = self.service.create(
            level,
            parent_id,
            {
                "name": name,
                "description": payload["description"],
                "businessOwner": payload["businessOwner"],
                "technicalOwner": tech,
            },
            actor=actor,
        )
        summary["created"] += 1
        return created["id"]


def _parse_row(row: tuple, get) -> dict:  # noqa: ANN001
    def node(n: str, d: str, b: str, t: str) -> Optional[dict]:
        name, description = get(row, n), get(row, d)
        business_owner, technical_owner = get(row, b), get(row, t)
        if not name and not description and not business_owner and not technical_owner:
            return None
        return {
            "name": name,
            "description": description,
            "businessOwner": business_owner,
            "technicalOwner": technical_owner,
        }

    return {
        "dg": node("Domain Group Name", "Domain Group Description", "Domain Group Business Owner", "Domain Group Technical Owner"),
        "domain": node("Domain Name", "Domain Description", "Domain Business Owner", "Domain Technical Owner"),
        "sub": node("Subdomain Name", "Subdomain Description", "Subdomain Business Owner", "Subdomain Technical Owner"),
        "product": node("Data Product Name", "Data Product Description", "Data Product Business Owner", "Data Product Technical Owner"),
    }


def _validate_row_shape(p: dict) -> None:
    if p["product"] and not p["sub"]:
        raise RowError("Data Product present but Subdomain is missing")
    if p["sub"] and not p["domain"]:
        raise RowError("Subdomain present but Domain is missing")
    if p["domain"] and not p["dg"]:
        raise RowError("Domain present but Domain Group is missing")

File: server/src/excelService.ts
import ExcelJS from "exceljs";
import type { DB } from "./db.js";
import { TaxonomyService } from "./taxonomyService.js";
import { Level, ImportSummary, ImportError, CreateInput } from "./types.js";
import { validateCreate, ValidationError, sameName, normalizeTechnicalOwner } from "./validation.js";

const SHEET = "Taxonomy";

// Column order. "Domain Group Technical Owner" is optional (DG tech owner is
// not required) but included so optional values round-trip cleanly.
export const COLUMNS = [
  "Domain Group Name",
  "Domain Group Description",
  "Domain Group Business Owner",
  "Domain Group Technical Owner",
  "Domain Name",
  "Domain Description",
  "Domain Business Owner",
  "Domain Technical Owner",
  "Subdomain Name",
  "Subdomain Description",
  "Subdomain Business Owner",
  "Subdomain Technical Owner",
  "Data Product Name",
  "Data Product Description",
  "Data Product Business Owner",
  "Data Product Technical Owner",
] as const;

// Columns that MUST be present in an imported file (DG Technical Owner excluded).
const REQUIRED_COLUMNS = COLUMNS.filter((c) => c !== "Domain Group Technical Owner");

interface RowValues {
  dg: CreateInput | null;
  domain: CreateInput | null;
  sub: CreateInput | null;
  product: CreateInput | null;
}

export class ExcelService {
  constructor(
    private db: DB,
    private service: TaxonomyService,
  ) {}

  // ---- Export ----------------------------------------------------------

  async export(): Promise<Buffer> {
    const wb = new ExcelJS.Workbook();
    wb.creator = "Data Mesh Taxonomy";
    wb.created = new Date();
    const ws = wb.addWorksheet(SHEET);
    ws.addRow([...COLUMNS]);
    ws.getRow(1).font = { bold: true };
    ws.views = [{ state: "frozen", ySplit: 1 }];
    COLUMNS.forEach((c, i) => {
      ws.getColumn(i + 1).width = Math.max(18, c.length + 2);
    });

    const tree = this.service.getTree();
    for (const g of tree) {
      const gCells = [g.name, g.description, g.businessOwner, g.technicalOwner ?? ""];
      if (g.children.length === 0) {
        ws.addRow([...gCells, "", "", "", "", "", "", "", "", "", "", "", ""]);
        continue;
      }
      for (const d of g.children) {
        const dCells = [d.name, d.description, d.businessOwner, d.technicalOwner ?? ""];
        if (d.children.length === 0) {
          ws.addRow([...gCells, ...dCells, "", "", "", "", "", "", "", ""]);
          continue;
        }
        for (const s of d.children) {
          const sCells = [s.name, s.description, s.businessOwner, s.technicalOwner ?? ""];
          if (s.children.length === 0) {
            ws.addRow([...gCells, ...dCells, ...sCells, "", "", "", ""]);
            continue;
          }
          for (const p of s.children) {
            ws.addRow([...gCells, ...dCells, ...sCells, p.name, p.description, p.businessOwner, p.technicalOwner ?? ""]);
          }
        }
      }
    }
    const buf = await wb.xlsx.writeBuffer();
    return Buffer.from(buf);
  }

  /** Build the blank import template (header + one example row). */
  async template(): Promise<Buffer> {
    const wb = new ExcelJS.Workbook();
    const ws = wb.addWorksheet(SHEET);
    ws.addRow([...COLUMNS]);
    ws.getRow(1).font = { bold: true };
    ws.addRow([
      "Customer", "Customer-facing data assets", "Jane Doe", "",
      "Loyalty", "Loyalty program data", "Jane Doe", "Raj Patel",
      "Membership", "Member profiles and tiers", "Jane Doe", "Raj Patel",
      "Member Profile", "Curated member profile product", "Jane Doe", "Raj Patel",
    ]);
    COLUMNS.forEach((c, i) => (ws.getColumn(i + 1).width = Math.max(18, c.length + 2)));
    const buf = await wb.xlsx.writeBuffer();
    return Buffer.from(buf);
  }

  // ---- Import ----------------------------------------------------------

  async import(buffer: Buffer): Promise<ImportSummary> {
    const wb = new ExcelJS.Workbook();
    await wb.xlsx.load(buffer);
    const ws = wb.getWorksheet(SHEET) ?? wb.worksheets[0];
    if (!ws) throw new ValidationError({ file: "Workbook contains no worksheets" });

    // Map header names -> column index (1-based).
    const headerRow = ws.getRow(1);
    const headerIndex: Record<string, number> = {};
    headerRow.eachCell((cell, col) => {
      headerIndex[cellText(cell.value).toLowerCase()] = col;
    });
    const missing = REQUIRED_COLUMNS.filter((c) => !(c.toLowerCase() in headerIndex));
    if (missing.length) {
      throw new ValidationError({ file: `Missing required column(s): ${missing.join(", ")}` });
    }

    const get = (row: ExcelJS.Row, name: string): string => {
      const col = headerIndex[name.toLowerCase()];
      return col ? cellText(row.getCell(col).value) : "";
    };

    const summary: ImportSummary = { created: 0, updated: 0, skipped: 0, errors: [] };

    // Caches: parent id keyed by hierarchy path, to dedupe across rows.
    const groupId = new Map<string, string>();
    const domainId = new Map<string, string>();
    const subId = new Map<string, string>();

    // Everything runs in one transaction: an unexpected throw rolls the whole
    // import back so existing data is never partially corrupted. Expected,
    // per-row validation problems are recorded and skipped without aborting.
    const run = this.db.transaction(() => {
      ws.eachRow((row, rowNumber) => {
        if (rowNumber === 1) return; // header
        const parsed = parseRow(row, get);

        // Entirely blank row -> skip silently.
        if (!parsed.dg && !parsed.domain && !parsed.sub && !parsed.product) {
          summary.skipped++;
          return;
        }

        try {
          validateRowShape(parsed, rowNumber);

          // Validate each present level up front so the row is all-or-nothing.
          if (parsed.dg) validateCreate("DomainGroup", parsed.dg);
          if (parsed.domain) validateCreate("Domain", parsed.domain);
          if (parsed.sub) validateCreate("Subdomain", parsed.sub);
          if (parsed.product) validateCreate("DataProduct", parsed.product);

          // Apply top-down, upserting by hierarchy-aware name key.
          const gKey = parsed.dg!.name.trim().toLowerCase();
          const gid = this.upsert("DomainGroup", null, parsed.dg!, groupId, gKey, summary);

          if (!parsed.domain) return;
          const dKey = `${gid}::${parsed.domain.name.trim().toLowerCase()}`;
          const did = this.upsert("Domain", gid, parsed.domain, domainId, dKey, summary);

          if (!parsed.sub) return;
          const sKey = `${did}::${parsed.sub.name.trim().toLowerCase()}`;
          const sid = this.upsert("Subdomain", did, parsed.sub, subId, sKey, summary);

          if (!parsed.product) return;
          this.upsert("DataProduct", sid, parsed.product, new Map(), "", summary);
        } catch (err) {
          if (err instanceof ValidationError) {
            const msg = Object.entries(err.fields)
              .map(([k, v]) => (k === "_" || k === "file" ? v : `${v}`))
              .join("; ");
            summary.errors.push({ row: rowNumber, message: msg });
          } else if (err instanceof RowError) {
            summary.errors.push({ row: rowNumber, message: err.message });
          } else {
            throw err; // unexpected -> abort & rollback everything
          }
        }
      });
    });

    run();
    return summary;
  }

  /** Find-or-create/update a node by name under a parent. Updates counters. */
  private upsert(
    level: Level,
    parentId: string | null,
    input: CreateInput,
    cache: Map<string, string>,
    cacheKey: string,
    summary: ImportSummary,
  ): string {
    const tech = normalizeTechnicalOwner(input.technicalOwner);
    const existing = this.service.list(level, parentId).find((n) => sameName(n.name, input.name));

    if (existing) {
      const changed =
        existing.description !== input.description.trim() ||
        existing.businessOwner !== input.businessOwner.trim() ||
        (existing.technicalOwner ?? null) !== tech ||
        existing.name !== input.name.trim();
      if (changed) {
        this.service.update(level, existing.id, {
          name: input.name,
          description: input.description,
          businessOwner: input.businessOwner,
          technicalOwner: tech,
        });
        summary.updated++;
      }
      if (cacheKey) cache.set(cacheKey, existing.id);
      return existing.id;
    }

    const created = this.service.create(level, parentId, { ...input, technicalOwner: tech });
    summary.created++;
    if (cacheKey) cache.set(cacheKey, created.id);
    return created.id;
  }
}

class RowError extends Error {}

function cellText(value: ExcelJS.CellValue): string {
  if (value == null) return "";
  if (typeof value === "object") {
    const anyVal = value as { text?: string; result?: unknown; richText?: { text: string }[] };
    if (Array.isArray(anyVal.richText)) return anyVal.richText.map((r) => r.text).join("").trim();
    if (typeof anyVal.text === "string") return anyVal.text.trim();
    if (anyVal.result != null) return String(anyVal.result).trim();
    return "";
  }
  return String(value).trim();
}

function parseRow(row: ExcelJS.Row, get: (r: ExcelJS.Row, name: string) => string): RowValues {
  const node = (n: string, d: string, b: string, t: string): CreateInput | null => {
    const name = get(row, n);
    const description = get(row, d);
    const businessOwner = get(row, b);
    const technicalOwner = get(row, t);
    if (!name && !description && !businessOwner && !technicalOwner) return null;
    return { name, description, businessOwner, technicalOwner };
  };
  return {
    dg: node("Domain Group Name", "Domain Group Description", "Domain Group Business Owner", "Domain Group Technical Owner"),
    domain: node("Domain Name", "Domain Description", "Domain Business Owner", "Domain Technical Owner"),
    sub: node("Subdomain Name", "Subdomain Description", "Subdomain Business Owner", "Subdomain Technical Owner"),
    product: node("Data Product Name", "Data Product Description", "Data Product Business Owner", "Data Product Technical Owner"),
  };
}

/** Ensure no level is present without its ancestors. */
function validateRowShape(p: RowValues, _row: number): void {
  if (p.product && !p.sub) throw new RowError("Data Product present but Subdomain is missing");
  if (p.sub && !p.domain) throw new RowError("Subdomain present but Domain is missing");
  if (p.domain && !p.dg) throw new RowError("Domain present but Domain Group is missing");
}
